source/common/upstream/cluster_manager_impl.cc

```diff
@@ -343,6 +372,80 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   }
 }
 
+bool ClusterManagerImpl::scheduleUpdate(const Cluster& cluster, uint32_t priority, bool mergeable) {
```

Comment:
Question: if, for a given cluster/priority, we have something like "add, remove, inplace update, add, remove, inplace update", should we be canceling the outstanding deferred update at the second "add"?

Reply:
Correct, because the second "add" is a non-mergeable update so it will trigger immediate delivery, thus  cancelling the pending merged update by disabling the timer.

This happens here in `scheduleUpdate()`:

```
  // Has an update_merge_window gone by since the last update? If so, don't schedule                                                                                              
  // the update so it can be applied immediately. Ditto if this is not a mergeable update.                                                                                        
  const auto delta = std::chrono::steady_clock::now() - updates->last_updated_;                                                                                                   
  const uint64_t delta_ms = std::chrono::duration_cast<std::chrono::milliseconds>(delta).count();                                                                                 
  if (delta_ms > timeout || !mergeable) {                                                                                                                                         
    // If there was a pending update, we count this update as merged update.                                                                                                      
    if (updates->disableTimer()) {                                                                                                                                                
      cm_stats_.merged_updates_.inc();                                                                                                                                            
    }                                                                                                                                                                             
                                                                                                                                                                                  
    updates->last_updated_ = std::chrono::steady_clock::now();                                                                                                                    
    return false;                                                                                                                                                                 
  } 
```
And we exercise a similar variation of this at the end at the unit test (inplace/inplace/inplace/remove→ timer disabled and updates delivered immediately).